cmake/options.cmake

```diff
@@ -146,6 +146,7 @@ set(DNNL_ENABLE_PRIMITIVE_GPU_ISA "ALL" CACHE STRING
     at build time. Regardless of value chosen, reference OpenCL-based
     implementations will always be available. Valid values:
     - ALL (the default). Includes all ISA to be enabled.
+    - NONE. Includes no ISAs, just the generic kernels.
```

Comment:
(random spot)

Thank you for taking the effort to enable a generic vendor for the OpenCL runtime. When it comes to external contributions that impact some aspects of the library's architecture a formal proposal (RFC) is required. Please make yourself familiar with the [contribution guidelines](https://github.com/oneapi-src/oneDNN?tab=readme-ov-file#contributing) and the [RFC process](https://github.com/oneapi-src/oneDNN/tree/rfcs).

While going through the changes in this pull-request I found a few things that have to be addressed. Your proposal is expected to cover it.
* According to the [GPU code organization structure](https://github.com/oneapi-src/oneDNN/tree/main/src/gpu#readme) the generic OpenCL kernels should reside in `gpu/generic/ocl`.
* The OpenCL primitive implementations that you marked as generic may be generic at this point but there is absolutely no guarantee that it will stay that way. The right way to make them generic is to move them to `gpu/generic/ocl` and enable them for the Intel and Generic vendors.
* The library architecture requires that one engine per runtime (CPU and GPU) must be enabled. Currently, there is only one OpenCL engine that is Intel specific. When `DNNL_GPU_VENDOR` is `GENERIC` and `DNNL_GPU_RUNTIME` is `OCL` there are no engines that could be used therefore a new engine for the generic vendor and OpenCL runtime has to be introduced.
* There shouldn't any vendor specific compile time checks in a vendor specific space. All code that resides in `gpu/intel` assumes that  `DNNL_GPU_VENDOR` is `INTEL`.
*  When `DNNL_GPU_VENDOR` is `INTEL` and `DNNL_GPU_RUNTIME` is `SYCL` the `xpu/ocl` code must be enabled (the new condition that you introduced seems to break the rule).
* I think the `DNNL_ENABLE_PRIMITIVE_GPU_ISA` option should be enabled (and defined) only for the Intel vendor and should be ignored otherwise
* USM is not part of the OpenCL standard and therefore we cannot assume that all OpenCL vendors support it. The USM utility functions must be conditionally enabled for the vendors that has the support.

As for enabling OpenCL kernels for CPU, we don't have any plans to do that and the library doesn't have any architecture to do that. If you are interested in that you may come up with an architecture design and publish a proposal.

And the last thing, we plan to enable the generic vendor for SYCL runtime in the coming months. As an option you could wait to see what the design will look like to get a better understanding of what it should look like for OpenCL.